HADOOP-17520. Downgrade missing DDB table to unguarded client

steveloughran opened the pull request:
If an s3a client starts up and cannot find/read the DDB table then it downgrades to being unguarded

steveloughran committed:
HADOOP-17520. Downgrade missing DDB table to unguarded client

Change-Id: If1411d2e6df2949b20c5dbd12b033703d8c8a6bd

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   2m 34s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |   |   0m  0s | [test4tests](test4tests) |  The patch appears to include 4 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  37m 26s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 41s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |   0m 34s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  checkstyle  |   0m 26s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 42s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 53s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   0m 29s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +0 :ok: |  spotbugs  |   1m  9s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m  5s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 33s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 37s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javac  |   0m 37s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 30s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  javac  |   0m 30s |  |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 19s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 34s |  |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 26s |  |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 17s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   0m 24s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| -1 :x: |  findbugs  |   1m  9s | [/new-findbugs-hadoop-tools_hadoop-aws.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/1/artifact/out/new-findbugs-hadoop-tools_hadoop-aws.html) |  hadoop-tools/hadoop-aws generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 51s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 29s |  |  The patch does not generate ASF License warnings.  |
|  |   |  85m 21s |  |  |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-tools/hadoop-aws |
|  |  Nullcheck of S3AInstrumentation.metricsSystem at line 688 of value previously dereferenced in org.apache.hadoop.fs.s3a.S3AInstrumentation.close()  At S3AInstrumentation.java:688 of value previously dereferenced in org.apache.hadoop.fs.s3a.S3AInstrumentation.close()  At S3AInstrumentation.java:[line 685] |
|  |  instanceof will always return false in org.apache.hadoop.fs.s3a.s3guard.S3Guard.getMetadataStore(FileSystem, ITtlTimeProvider), since a RuntimeException can't be a java.io.IOException  At S3Guard.java:in org.apache.hadoop.fs.s3a.s3guard.S3Guard.getMetadataStore(FileSystem, ITtlTimeProvider), since a RuntimeException can't be a java.io.IOException  At S3Guard.java:[line 137] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2689 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 55841faba7c7 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 0e2b3086e3b |
| Default Java | Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/1/testReport/ |
| Max. process+thread count | 597 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/1/console |
| versions | git=2.25.1 maven=3.6.3 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-17520. Downgrading S3Guard: the test experience.

* report on downgrade
* treat AccessDeniedException as downgrade (surfaces in RoleDelegationTokens)
* close (failing) metastore on exception -reduces ref count of client creds.
* test base has static field to cache whether S3Guard is downgrading
* which allows test suites/parameterized runs which need S3Guard to skip
  in test setup before creating an S3AFS. Reduces test case execution time to
  a few tens of millis, rather than seconds.
* adopt in tests where appropriate
* improve test cleanup in places where fs==null in teardown caused problems.

TODO: docs.
It's still better to explicitly disable as you will save the overhead of talking
to DDB only to find the table isn't there.

Change-Id: I1d01a59dd736b232673a798ce52c846fb451fb8c

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m 24s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |   |   0m  0s | [test4tests](test4tests) |  The patch appears to include 17 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  35m 33s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 41s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |   0m 35s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  checkstyle  |   0m 26s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 40s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 37s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 21s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   0m 28s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +0 :ok: |  spotbugs  |   1m  6s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m  4s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 33s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 36s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javac  |   0m 36s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 29s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  javac  |   0m 29s |  |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 19s | [/diff-checkstyle-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt) |  hadoop-tools/hadoop-aws: The patch generated 4 new + 40 unchanged - 0 fixed = 44 total (was 40)  |
| +1 :green_heart: |  mvnsite  |   0m 33s |  |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 21s |  |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 16s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   0m 25s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| -1 :x: |  findbugs  |   1m 10s | [/new-findbugs-hadoop-tools_hadoop-aws.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/2/artifact/out/new-findbugs-hadoop-tools_hadoop-aws.html) |  hadoop-tools/hadoop-aws generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 47s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  |  The patch does not generate ASF License warnings.  |
|  |   |  81m 44s |  |  |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-tools/hadoop-aws |
|  |  Nullcheck of S3AInstrumentation.metricsSystem at line 688 of value previously dereferenced in org.apache.hadoop.fs.s3a.S3AInstrumentation.close()  At S3AInstrumentation.java:688 of value previously dereferenced in org.apache.hadoop.fs.s3a.S3AInstrumentation.close()  At S3AInstrumentation.java:[line 685] |
|  |  instanceof will always return false in org.apache.hadoop.fs.s3a.s3guard.S3Guard.getMetadataStore(FileSystem, ITtlTimeProvider), since a RuntimeException can't be a java.io.IOException  At S3Guard.java:in org.apache.hadoop.fs.s3a.s3guard.S3Guard.getMetadataStore(FileSystem, ITtlTimeProvider), since a RuntimeException can't be a java.io.IOException  At S3Guard.java:[line 139] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2689 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c945e796bfe2 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 9434c1eccc2 |
| Default Java | Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/2/testReport/ |
| Max. process+thread count | 555 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2689/2/console |
| versions | git=2.25.1 maven=3.6.3 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
+docs to emphasise that it's still best to disable s3guard explicitly

bogthe commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestS3GuardConcurrentOps.java:
```diff
@@ -65,6 +70,12 @@ protected Configuration createConfiguration() {
     return conf;
   }
 
+  @Override
+  public void setup() throws Exception {
+    super.setup();
+    assumeFilesystemHasMetadatastore(getFileSystem());
+  }
+
```
With the new checks added in `AbstractS3ATestBase` this looks obsolete since the same logic is happening in parent `setup` method:
```
    if (isS3GuardRequired()) {
      downgradeIfUnguarded(fs);
    }
``` 

Not that important tho since hopefully s3 guard will be removed 🤞

bogthe reviewed (approved):
I like this. Added benefit for newcomers to the code base is that it makes it more explicit which tests require s3guard or not.

steveloughran commented:
closing as we are removing s3guard completely
